### PR TITLE
Simple evaluation of FunctionWrapper functions in python

### LINKED ISSUE #20420: Implement Evaluate for FunctionWrapper
Implement Evaluate for the new FunctionWrapper python class introduced by issue #19010 as described in the design document https://github.com/mantidproject/documents/blob/master/Design/Fit%20functions%20in%20python.md 

Currently evaluation can only be done of workspaces as in ``FitFunctionsTest`` ``test_evaluation`` and ``test_arithmetic.``

### PR DESCRIPTION
**Pull request by KarlPalmen**
Fixes #20420 

**To test**
Try python scripts such as
```
p = Polynomial(n=2, A0=0, A1=0.5, A2=0.5)

print p(1)
print p(2)
print p(3)
print p([0,1,2,3])

ws = CreateWorkspace(DataX=[0,1,2,3,4,5,6,7], DataY=[5,5,5,5,5,5,5])

print p(ws).readY(1)
```
Also try with numpy arrays.

**Release Notes** 
Linked through existing release note.

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

File: Framework/PythonInterface/test/python/mantid/FitFunctionsTest.py [hunk 1]
```diff
@@ -574,8 +574,47 @@ def test_arithmetic(self):
...
+
+    def test_evaluation_by_2D_numpy_array(self):
+        import numpy as np
+        a = np.array([[0, 1],[2, 3]])
+        p = Polynomial(n=4, A0=1, A1=1, A2=1, A3=1, A4=1)
+        result = p(a)
+        self.assertAlmostEqual(result[0][0],1.0)
+        self.assertAlmostEqual(result[0][1],5.0)
+        self.assertAlmostEqual(result[1][0],31.0)
+        self.assertAlmostEqual(result[1][1],121.0)
```
Comment (mantid-roman) [inline Framework/PythonInterface/test/python/mantid/FitFunctionsTest.py]: A test is needed that `*params` arguments set the parameters correctly

**Comment by KarlPalmen** [comment]
I've realised I need to add something to the user documentation about this. I'll do this after getting Roman's change on my local copy of the branch.

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 2]
```diff
@@ -109,6 +109,50 @@ def __mul__ (self, other):
...
       return prod
+      
+  def __call__(self, x, *params):
+      """ Implement function evaluation, such that
+          func(args) is equivalent func.__call__(args)
+           
+          :param x: x value or list of x values
+          :param *params list of parameter values
+      """
+      from mantid.simpleapi import CreateWorkspace, EvaluateFunction
```
Comment (mantid-roman) [inline Framework/PythonInterface/mantid/fitfunctions.py]: There is no need for these imports any more. I forgot to delete the line after my changes.

**Review by mantid-roman — CHANGES_REQUESTED** [review]

**Comment by mantid-roman** [comment]
@KarlPalmen I made some changes to make it possible to pass function objects to fitting algorithms. Now mantid functions can be used with the `scipy` optimisation library.

**Comment by KarlPalmen** [comment]
@mantid-roman I've tried SimpleAPIRenameWorkspaceTest on my PC and it crashes.

**Comment by KarlPalmen** [comment]
Thank you @mantid-roman . That fixed it.

**Comment by KarlPalmen** [comment]
I'm un-assigning @mantid-roman from the PR because he has been doing work on this issue. Thank you. Someone else needs to test & review it.

File: Framework/PythonInterface/mantid/fitfunctions.py [hunk 3]
```diff
@@ -109,6 +109,49 @@ def __mul__ (self, other):
...
+          :param x: x value or list of x values
+          :param *params list of parameter values
+      """
+      import numpy as np
+      
+      if isinstance(x, Workspace):
+          # If the input is a workspace, simply return the output workspace.
+          return self._execute_algorithm('EvaluateFunction', Function=self.fun, InputWorkspace=x)
+       
+      list_input = False
```
Comment (josephframsay) [inline Framework/PythonInterface/mantid/fitfunctions.py]: These two flags might read more intuitively if they were written something like `list_input = isinstance(x, list)`

**Review by josephframsay — CHANGES_REQUESTED** [review]
Changes look good, but there's a few places where, I might be wrong, they don't appear to quite meet the design document:

- **"A call without passing in any initialisation arguments should create a function with default parameters and attributes"** - this is true of most of the functions, but if I make a `Polynomial` with no parameters and then try and call it, I get `IndexError: ParamFunction parameter index 0 out of range 0`. If it's not valid to have a `Polynomial` with no parameters, maybe there should be a check for this (though maybe that's a problem with `Polynomial` and not your new stuff)?
- **"The output type is the same as that of the input"** - This is not the case if I call functions with a list, as I always get back an array. Is this deliberate?

**Comment by mantid-roman** [comment]
@josephframsay 
* It is a problem with `Polynomial`
* I think returning a numpy array for a list is fine.

**Review by mantid-roman — APPROVED** [review]

**Comment by KarlPalmen** [comment]
@josephframsay  have you any outstanding concerns about this pull request?

**Comment by josephframsay** [comment]
Only a superficial thing - I think the lines associated with `list_input` and `numpy_input` could be made a bit more readable by changing them to something like:
```
list_input = isinstance(x, list)
numpy_input = isinstance(x, np.ndarray)
if numpy_input:
    ....
elif list_input:
    ....
```
It means the variables only get assigned to once, and it might save someone a bit of a headache in future. Other than that, everything looks grand to me

**Review by josephframsay — APPROVED** [review]

**Pull request merged**